app/controllers/algorithms_controller.rb

```diff
@@ -0,0 +1,45 @@
+class AlgorithmsController < ApplicationController
+  before_action :authenticate_user!
+  SUPER_EXPRESSION = /\s*def(.|\n)*?end\s*/
+
+  def show
+    @secret = Secret.find_by(id: params[:secret_id])
+    @algorithm = Algorithm.find(params[:id])
+  end
+
+  def run_code
+    @secret = Secret.find_by(id: params[:secret_id])
```

Comment:
I'd refactor this into a module / somewhere that's not a big gnarly controller method.